include/ensmallen_bits/nsga2/nsga2.hpp

```diff
@@ -0,0 +1,274 @@
+/**
+ * @file nsga2.hpp
+ * @author Sayan Goswami
+ *
+ * NSGA-II is a multi-objective optimzation algorithm, widely used in
+ * many real-world applications. NSGA-II genererates offsprings using
+ * crossover and mutation and then selects the next generation according
+ * to nondominated-sorting and crowding distance comparison.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef ENSMALLEN_NSGA2_NSGA2_HPP
+#define ENSMALLEN_NSGA2_NSGA2_HPP
+
+namespace ens {
+
+/**
+ * NSGA-II (Non-dominated Sorting Genetic Algorithm - II) is a multi-objective
+ * optimzation algorithm. This class implements the NSGA-II algorithm.
+ *
+ * The algorithm works by generating a candidate population from a fixed
+ * starting point. At each stage of optimization, a new population of children
+ * is generated. This new population alongwith its predecessor is sorted using
+ * non-domiation as the metric. Following this, the population is further
+ * seggregated in fronts. A new population is generated from these fronts having
+ * size equal to that of the starting population.
+ *
+ * During evolution, two parents are randomly chosen using binary tournament
+ * selection. A pair of children are generated by crossing over these two
+ * candidates followed by mutation.
+ *
+ * The best front (pareto optimal) is returned by the Optimize() method.
+ *
+ * For more information, see the following:
+ *
+ * @code
+ * @article{10.1109/4235.996017,
```

Comment:
I think to keep this simple, we could just list the required fields: author, title, journal, year, we can keep the url as well.